sol-anchor/idl.go

```diff
@@ -0,0 +1,612 @@
+package solanchor
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"unicode"
+
+	"github.com/golang-cz/textcase"
+)
+
+type IDL struct {
+	Events       []Event       `json:"events"`
+	Instructions []Instruction `json:"instructions"`
+	Metadata     Metadata      `json:"metadata"`
+	Types        []Type        `json:"types"`
+}
+
+type Metadata struct {
+	Address string `json:"address"`
+}
+
+// --- EVENTS
+
+type Event struct {
+	Name   string  `json:"name"`
+	Fields []Field `json:"fields"`
+}
+
+func (e *Event) SnakeCaseName() string {
+	return toSnakeCase(e.Name, true)
+}
+
+// --- INSTRUCTIONS
+type Instruction struct {
+	Name string  `json:"name"`
+	Args []Field `json:"args"`
+}
+
+func (i *Instruction) PascalCaseName() string {
+	return textcase.PascalCase(i.Name)
+}
+
+func (i *Instruction) SnakeCaseName() string {
+	return toSnakeCase(i.Name, true)
+}
+
+// --- FIELDS
+
+type Field struct {
+	Name string    `json:"name"`
+	Type FieldType `json:"type"`
+	//Index bool      `json:"index"`
+}
+
+func (f *Field) SnakeCaseName() string {
+	return toSnakeCase(f.Name, true)
+}
+
+func (f *Field) SnakeCaseNameWithoutInitialUnderscore() string {
+	return toSnakeCase(f.Name, false)
+}
+
+// FIELD TYPE
+
+type FieldTypeArray struct {
+	Type   string
+	Length string
+}
+
+type FieldTypeVec struct {
+	Type      string
+	IsDefined bool
+}
+
+type FieldTypeOption struct {
+	Simple  *string
+	Defined *string
+	Vec     *FieldTypeVec
+	Array   *FieldTypeArray
+}
+
+func (f *FieldTypeOption) IsSimple() bool {
+	return f.Simple != nil
+}
+
+func (f *FieldTypeOption) IsDefined() bool {
+	return f.Defined != nil
+}
+
+func (f *FieldTypeOption) IsVec() bool {
+	return f.Vec != nil
+}
+
+func (f *FieldTypeOption) IsArray() bool {
+	return f.Array != nil
+}
+
+type FieldType struct {
+	Simple  string
+	Defined string
+	Array   *FieldTypeArray
+	Vec     *FieldTypeVec
+	Option  *FieldTypeOption
+}
+
+func (t *FieldType) IsSimple() bool {
+	return t.Simple != ""
+}
+
+func (t *FieldType) IsSimplePubKey() bool {
+	return t.Simple == "publicKey"
+}
+
+func (t *FieldType) IsDefined() bool {
+	return t.Defined != ""
+}
+
+func (t *FieldType) IsArray() bool {
+	return t.Array != nil
+}
+
+func (t *FieldType) IsVec() bool {
+	return t.Vec != nil
+}
+
+func (t *FieldType) IsOption() bool {
+	return t.Option != nil
+}
+
+func (t *FieldType) Resolve() string {
```

Comment:
not to start a flamewar here, but a `switch case` would reduce this function to quite a few less lines (if it fits in my code editor window, it's easier to read quickly)